xapian-core/include/xapian/cluster.h

```diff
@@ -180,64 +182,153 @@ class XAPIAN_VISIBILITY_DEFAULT TermListGroup : public FreqSource {
     void add_document(const Document &doc);
 
     // Add a number of documents from the DocumentSource
-    void add_documents(DocumentSource &docs);
+    void add_documents(MSetDocumentSource docs);
 
     Xapian::doccount get_termfreq(const std::string &tname);
 
     Xapian::doccount get_doccount();
 };
 
+/** Structure to store term and corresponding wdf.
+ *  This is used with PointTermIterator to return wdf
+ */
+struct XAPIAN_VISIBILITY_DEFAULT Wdf {
+    std::string term;
+    double wdf;
+    Wdf(std::string term_, double wdf_) : term(term_), wdf(wdf_) {}
+};
+
+// Represent a single point in the vector space
+class XAPIAN_VISIBILITY_DEFAULT Point {
+    std::vector<struct Wdf> termlist;
+    std::map<std::string, double> values;
+    Document doc;
+  public:
+    // Returns an iterator to beginning of termlist
+    TermIterator termlist_begin();
+
+    // Returns an iterator to end of termlist
+    TermIterator termlist_end();
+
+    // Checks whether the point contains the a certain term
+    bool contains(std::string term);
+
+    // Returns the TF-IDF weight of the required term
+    double get_value(std::string term);
+
+    // Divide the TF-IDF weights for all terms by a given value
+    void divide(int size);
+
+    // Initialize the point with its weights
+    void initialize(TermListGroup &tlg, const Document &doc);
+
+    // Add a value to the associated term
+    void add_value(std::string term, double value);
+
+    // Set a weight value for the associated term
+    void set_value(std::string term, double value);
+
+    // Returns the doucment associated with this Point
+    Document get_document();
+};
+
 // Base class for calculating the similarity between documents
 class XAPIAN_VISIBILITY_DEFAULT Similarity {
   public:
     // Destructor
     virtual ~Similarity();
 
     // Calculates the similarity between the two documents
-    virtual double similarity(TermListGroup tlg, TermIterator a_begin, TermIterator a_end, TermIterator b_begin, TermIterator b_end) = 0;
+    //virtual double similarity(TermIterator a_begin, TermIterator a_end, TermIterator b_begin, TermIterator b_end) = 0;
+    virtual double similarity(Point a, Point b) = 0;
 
    // Returns description of the similarity metric being used
     virtual std::string get_description() = 0;
 };
 
 // Class for caluclating the euclidian distance between two documents
 class XAPIAN_VISIBILITY_DEFAULT EuclidianDistance : public Similarity {
+    TermListGroup tlg;
   public:
-    double similarity(TermListGroup tlg, TermIterator a_begin, TermIterator a_end, TermIterator b_begin, TermIterator b_end);
+    EuclidianDistance(TermListGroup tlg_) : tlg(tlg_) {}
+
+    double similarity(Point a, Point b);
+
+    std::string get_description();
+};
+
+// Class for calculating the cosine distance between two documents
+class XAPIAN_VISIBILITY_DEFAULT CosineDistance : public Similarity {
+    TermListGroup tlg;
+  public:
+    CosineDistance(TermListGroup tlg_) : tlg(tlg_) {}
```

Comment:
Thanks for pointing this out! I forgot to change the constructor when I changed the way cosine  similarity function was implemented.